cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/DefaultArtifactManager.java

```diff
@@ -0,0 +1,197 @@
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .registerTypeAdapter(PluginClass.class, new PluginClassDeserializer())
+    .create();
+
+  private final DiscoveryServiceClient discoveryServiceClient;
+  private final File tmpDir;
+  private final ClassLoader bootstrapClassLoader;
+  private final LocationFactory locationFactory;
+
+  @Inject
+  public DefaultArtifactManager(CConfiguration cConf, DiscoveryServiceClient discoveryServiceClient,
+                                LocationFactory locationFactory) {
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.locationFactory = locationFactory;
+    File tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    this.tmpDir = DirUtils.createTempDir(tmpDir);
+    // There is no reliable way to get bootstrap ClassLoader from Java (System.class.getClassLoader() may return null).
+    // A URLClassLoader with no URLs and with a null parent will load class from bootstrap ClassLoader only.
+    this.bootstrapClassLoader = new URLClassLoader(new URL[0], null);
+  }
+
+  /**
+   * For the specified namespace, return the available artifacts in the namespace and also the artifacts in system
+   * namespace
+   *
+   * @param namespaceId namespace
+   * @return {@link List<ArtifactInfo>}
+   * @throws IOException If there are any exception while retrieving artifacts
+   */
+  public List<ArtifactInfo> listArtifacts(NamespaceId namespaceId) throws IOException {
+    URL url = discoverServiceAndCreateURL(String.format("namespaces/%s/artifact-internals/list/artifacts",
+                                                        namespaceId.getEntityName()));
+
+    HttpResponse httpResponse = HttpRequests.execute(HttpRequest.get(url).build());
+
+    if (httpResponse.getResponseCode() == HttpResponseStatus.NOT_FOUND.getCode()) {
+      throw new IOException("Could not list artifacts, endpoint not found");
+    }
+
+    if (isSuccessful(httpResponse.getResponseCode())) {
+      List<ArtifactInfo> artifactInfoList =
+        GSON.fromJson(httpResponse.getResponseBodyAsString(), new TypeToken<List<ArtifactInfo>>() { }.getType());
+      return artifactInfoList;
+    } else {
+      throw new IOException(String.format("Exception while getting artifacts list %s",
+                                          httpResponse.getResponseBodyAsString()));
+    }
+  }
+
+  private boolean isSuccessful(int responseCode) {
+    return responseCode == 200;
+  }
+
+  private URL discoverServiceAndCreateURL(String pathSuffix) throws IOException {
+    ServiceDiscovered discovered = discoveryServiceClient.discover(Constants.Service.APP_FABRIC_HTTP);
+    Discoverable discoverable = new RandomEndpointStrategy(discovered).pick(1, TimeUnit.SECONDS);
+
+    if (discoverable == null) {
+      throw new IOException(String.format("Could not discover %s Service", Constants.Service.APP_FABRIC_HTTP));
```

Comment:
this should be throwing a ServiceUnavailable so that we know it can be retried. This class can extend the RemoteClient class to avoid repeat code.